source/particle/generator/probability_density_function.cc

```diff
@@ -0,0 +1,295 @@
+            const std::map<double,types::LevelInd>::const_iterator select_cell_iterator = cells.lower_bound(random_weight);
+            if ((random_weight < start_weight) || (select_cell_iterator == cells.end()))
+              continue;
+
+            const types::LevelInd select_cell = select_cell_iterator->second;
+
+            const typename parallel::distributed::Triangulation<dim>::active_cell_iterator
+            it (&(this->get_triangulation()), select_cell.first, select_cell.second);
+
+            Point<dim> max_bounds, min_bounds;
+            // Get the bounds of the cell defined by the vertices
+            for (unsigned int d=0; d<dim; ++d)
+              {
+                min_bounds[d] = INFINITY;
+                max_bounds[d] = -INFINITY;
+              }
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell; ++v)
+              {
+                const Point<dim> vertex_position = it->vertex(v);
+                for (unsigned int d=0; d<dim; ++d)
+                  {
+                    min_bounds[d] = fmin(vertex_position[d], min_bounds[d]);
+                    max_bounds[d] = fmax(vertex_position[d], max_bounds[d]);
+                  }
+              }
+
+            // Generate random points in these bounds until one is within the cell
+            unsigned int num_tries = 0;
+            Point<dim> particle_position;
+            while (num_tries < 100)
+              {
+                for (unsigned int d=0; d<dim; ++d)
+                  {
+                    particle_position[d] = uniform_distribution_01(random_number_generator) *
+                                           (max_bounds[d]-min_bounds[d]) + min_bounds[d];
+                  }
+                try
+                  {
+                    const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(it, particle_position);
+                    if (GeometryInfo<dim>::is_inside_unit_cell(p_unit)) break;
+                  }
+                catch (typename Mapping<dim>::ExcTransformationFailed &)
+                  {
+                    // The point is not in this cell. Do nothing, just try again.
+                  }
+                num_tries++;
+              }
+            AssertThrow (num_tries < 100, ExcMessage ("Couldn't generate particle (unusual cell shape?)."));
+
+            // Add the generated particle to the set
+            const Particle<dim> new_particle(particle_position, cur_id);
+            particles.insert(std::make_pair(select_cell, new_particle));
+          }
+
+        return particles;
+      }
+
+
+
+      template <int dim>
+      void
+      ProbabilityDensityFunction<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Tracers");
+          {
+            prm.declare_entry ("Number of tracers", "1000",
+                               Patterns::Double (0),
```

Comment:
I don't think this parameter should be declared in the current file. It is a parameter not specific to this particular plugin. It should probably be declared by the particle manager.